In [1]:
import numpy as np

The function *gradientFunc()* is used for the calculation of the gradient of the given minimization function including the penalty element. The actual coeficients were checked using the software Wolfram Alpha.

In [2]:
def gradientFunc(x, mu):
    return np.matrix([[4*(x[0]-2)**3+2*(x[0]-2*x[1])-4*mu*x[0]*(x[1]-x[0]**2)],
                      [-4*(x[0]-2*x[1])+2*mu*(x[1]-x[0]**2)]])

Similarly, the function *hessianFunc()* is used for the calculation of the Hessian matrix of the given function.

In [3]:
def hessianFunc(x, mu):
    return np.matrix([[12*(mu+1)*x[0]**2 - 4*mu*x[1] - 48*x[0] + 50, -4-4*mu*x[0]],[-4-4*mu*x[0], 8+2*mu]])

Newton method follows the same steps as in the Problem Set 1, just with the addition of the gradual increase of the penalty parameter Mu. At each iteration of  the algorithm the parameter is multiplied by 1.1.

In [4]:
def newton_m(x0, e, mu):
    #calculate the norm of the gradient
    delta = np.linalg.norm(gradientFunc(x0, mu))
    #number of iterations
    i=0
    #while the norm is larger than tolerance level
    while delta > e:
        #multiply inverted hessian and gradient matrices, then modify them to required vector form
        #and subtract from previous point
        x0 = np.subtract(x0,np.matmul(np.linalg.inv(hessianFunc(x0, mu)), gradientFunc(x0, mu)).flatten().tolist()[0])
        #calculate the new norm
        delta = np.linalg.norm(gradientFunc(x0, mu))
        #increase the number of iterations
        i+=1
        #increase of the parameter mu
        mu*=1.1
    #print the results
    print ('Number of iterations is: {}'.format(i))
    print ('Root is at: {}'.format(x0))

In [5]:
newton_m([5, 5], 10**(-5), 10)

Number of iterations is: 66
Root is at: [0.94569071 0.89398738]


In [6]:
for mu in range(10, 100000, 10000):
    print(mu)
    newton_m([1, 1], 10**(-6), mu)
    print("===================")

10
Number of iterations is: 90
Root is at: [0.94559393 0.89411299]
10010
Number of iterations is: 18
Root is at: [0.94559344 0.89411364]
20010
Number of iterations is: 6
Root is at: [0.94559937 0.89410588]
30010
Number of iterations is: 6
Root is at: [0.94559391 0.89411296]
40010
Number of iterations is: 6
Root is at: [0.94559117 0.89411651]
50010
Number of iterations is: 6
Root is at: [0.94558953 0.89411863]
60010
Number of iterations is: 6
Root is at: [0.94558843 0.89412005]
70010
Number of iterations is: 6
Root is at: [0.94558765 0.89412106]
80010
Number of iterations is: 6
Root is at: [0.94558706 0.89412181]
90010
Number of iterations is: 6
Root is at: [0.94558661 0.8941224 ]


In [7]:
for mu in range(10, 100000, 10000):
    print(mu)
    newton_m([10000, 10000], 10**(-6), mu)
    print("===================")

10
Number of iterations is: 90
Root is at: [0.94559393 0.89411299]
10010
Number of iterations is: 57
Root is at: [0.94558325 0.89412688]
20010
Number of iterations is: 58
Root is at: [0.94558311 0.89412705]
30010
Number of iterations is: 58
Root is at: [0.94558307 0.8941271 ]
40010
Number of iterations is: 58
Root is at: [0.94558305 0.89412712]
50010
Number of iterations is: 58
Root is at: [0.94558304 0.89412714]
60010
Number of iterations is: 58
Root is at: [0.94558303 0.89412715]
70010
Number of iterations is: 58
Root is at: [0.94558303 0.89412715]
80010
Number of iterations is: 58
Root is at: [0.94558302 0.89412716]
90010
Number of iterations is: 58
Root is at: [0.94558302 0.89412716]


As we can see by testing the algorithm with several different starting points and values of Mu, the algorithm still converges to approximately $x_1=0.945$ and $x_2=0.894$. The only difference is, that the further away the starting point is - the larger amount of iterations is necessary to converge. Additionally, if the parameter Mu is too little the necessary number of iterations will be larger, but we still have to make sure it is not too large so it does not create convergence difficulties for the Newton method.